src/psim/core/configuration.cpp

```diff
@@ -0,0 +1,147 @@
+//
+// MIT License
+//
+// Copyright (c) 2020 Pathfinder for Autonomous Navigation (PAN)
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+//
+
+/** @file psim/core/configuration.cpp
+ *  @author Kyle Krol
+ */
+
+#include <psim/core/configuration.hpp>
+
+#include <fstream>
+#include <regex>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace psim {
+
+template <typename T>
+void Configuration::add(std::string const &name, T &&value) {
+  auto const iter = _parameters.find(name);
+  if (iter != _parameters.end())
+    throw std::runtime_error("Two or more parameters under the name: " + name);
+
+  _parameters[name] = new Parameter<T>(std::forward<T>(value));
+}
+
+Configuration::Configuration(Configuration &&config)
+    : _parameters(std::move(config._parameters)) { }
+
+Configuration &Configuration::operator=(Configuration &&config) {
+  _parameters = std::move(config._parameters);
+  return *this;
+}
+
+Configuration::~Configuration() {
+  for (auto &parameter : _parameters)
+    delete parameter.second;
+}
+
+ParameterBase const &Configuration::operator[](std::string const &name) const {
+  auto const *parameter_ptr = this->get(name);
+  if (!parameter_ptr)
+    throw std::runtime_error("Parameter not found with name: " + name);
+
+  return *parameter_ptr;
+}
+
+ParameterBase const *Configuration::get(std::string const &name) const {
+  auto const iter = _parameters.find(name);
+  return (iter == _parameters.end() ? nullptr : iter->second);
+}
+
+Configuration Configuration::make(std::string const &file) {
```

Comment:
Here I would've considered writing a parser using existing frameworks like GNU bison. Thankfully the parsing is simpler enough right now but if we discover later on that we want more options, such as macros to keep duplicating configurations, it will make our lives a million times easier.

Reply:
Agreed - to start of with however I'd just keep it as is. I don't think there's a need to make it more complicated or the time being.